File: modules/common/src/main/java/com/sygic/modules/common/di/ModulesComponent.java
package com.sygic.modules.common.di;

import android.app.Application;
import com.sygic.modules.common.di.module.*;
import com.sygic.modules.common.initialization.manager.SdkInitializationManager;
import com.sygic.modules.common.mapinteraction.manager.MapInteractionManager;
import com.sygic.modules.common.poi.manager.PoiDataManager;
import com.sygic.modules.common.theme.ThemeManager;
import com.sygic.ui.common.sdk.location.LocationManager;
import com.sygic.ui.common.sdk.model.ExtendedCameraModel;
import com.sygic.ui.common.sdk.model.ExtendedMapDataModel;
import com.sygic.ui.common.sdk.permission.PermissionsManager;
import dagger.Component;

import javax.inject.Singleton;

@Singleton
@Component(
        modules = {
                AppModule.class,
                MapModule.class,
                PoiDataManagerModule.class,
                SdkInitializationManagerModule.class,
                PermissionsModule.class,
                LocationModule.class,
                ThemeModule.class
        }
)
public interface ModulesComponent {
    ExtendedMapDataModel getMapDataModel();
    ExtendedCameraModel getCameraModel();
    MapInteractionManager getMapInteractionManager();
    PoiDataManager getPoiDataManager();
    SdkInitializationManager getSdkInitializationManager();
    PermissionsManager getPermissionsManager();
    LocationManager getLocationManager();
    ThemeManager getThemeManager();
    Application getApplication();
}

File: modules/common/src/main/java/com/sygic/modules/common/di/ModulesComponent.kt
package com.sygic.modules.common.di

import android.content.Context
import com.sygic.modules.common.initialization.manager.SdkInitializationManager
import com.sygic.modules.common.mapinteraction.manager.MapInteractionManager
import com.sygic.modules.common.poi.manager.PoiDataManager
import com.sygic.ui.common.sdk.model.ExtendedMapDataModel
import dagger.Component
import javax.inject.Singleton

@Singleton
@Component(
    modules = [
        ContextModule::class,
        MapDataModelModule::class,
        PoiDataManagerModule::class,
        MapInteractionManagerModule::class,
        SdkInitializationManagerModule::class
    ]
)
interface ModulesComponent {
    val context: Context
    val poiDataManager: PoiDataManager
    val extendedMapDataModel: ExtendedMapDataModel
    val mapInteractionManager: MapInteractionManager
    val sdkInitializationManager: SdkInitializationManager
}

interface ModuleBuilder<out T> {
    fun plus(component: ModulesComponent): ModuleBuilder<T>
    fun build(): T
}